# An Introduction to Hail

Before we can get started with the genomic analysis, we need to first upgrade Databrick's version of Python to 3.6. We'll do so using Anaconda.

First, download the Anaconda3 install script and place it in `/dbfs/tmp/` using the below command.

In [3]:
%sh sudo wget https://repo.anaconda.com/archive/Anaconda3-5.2.0-Linux-x86_64.sh -O /dbfs/tmp/Anaconda3-5.2.0-Linux-x86_64.sh

--2018-10-24 15:21:47-- https://repo.anaconda.com/archive/Anaconda3-5.2.0-Linux-x86_64.sh
Resolving repo.anaconda.com (repo.anaconda.com)... 104.17.108.77, 104.17.109.77, 104.17.110.77, ...
Connecting to repo.anaconda.com (repo.anaconda.com)|104.17.108.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 651745206 (622M) [application/x-sh]
Saving to: ‘/dbfs/tmp/Anaconda3-5.2.0-Linux-x86_64.sh’

 0K .......... .......... .......... .......... .......... 0% 3.80M 2m43s
 50K .......... .......... .......... .......... .......... 0% 3.32M 2m55s
 100K .......... .......... .......... .......... .......... 0% 2.96M 3m7s
 150K .......... .......... .......... .......... .......... 0% 3.31M 3m7s
 200K .......... .......... .......... .......... .......... 0% 3.98M 3m1s
 250K .......... .......... .......... .......... .......... 0% 3.95M 2m57s
 300K .......... .......... .......... .......... .......... 0% 4.00M 2m54s
 350K .......... .......... .......... .......... .......... 0% 3.58M 2m54s
 400K .......... .......... .......... .......... .......... 0% 4.12M 2m51s
 450K .......... .......... .......... .......... .......... 0% 3.72M 2m51s
 500K .......... .......... .......... .......... .......... 0% 4.04M 2m49s
 550K .......... .......... .......... .......... .......... 0% 3.47M 2m50s
 600K .......... .......... .......... .......... .......... 0% 4.10M 2m49s
 650K .......... .......... .......... .......... .......... 0% 3.77M 2m48s
 700K .......... .......... .......... .......... .......... 0% 3.77M 2m48s
 750K .......... .......... .......... .......... .......... 0% 3.10M 2m50s
 800K .......... .......... .......... .......... .......... 0% 3.76M 2m50s
 850K .......... .......... .......... .......... .......... 0% 3.90M 2m49s
 900K .......... .......... .......... .......... .......... 0% 3.93M 2m48s
 950K .......... .......... .......... .......... .......... 0% 3.42M 2m49s
 1000K .......... .......... .......... .......... .......... 0% 3.90M 2m49s
 1050K .......... .......... .......... .......... .......... 0% 4.36M 2m47s
 1100K .......... .......... .......... .......... .......... 0% 4.06M 2m47s
 1150K .......... .......... .......... .......... .......... 0% 3.49M 2m47s
 1200K .......... .......... .......... .......... .......... 0% 4.08M 2m47s
 1250K .......... .......... .......... .......... .......... 0% 4.06M 2m46s
 1300K .......... .......... .......... .......... .......... 0% 4.16M 2m45s
 1350K .......... .......... .......... .......... .......... 0% 3.51M 2m46s
 1400K .......... .......... .......... .......... .......... 0% 4.13M 2m45s
 1450K .......... .......... .......... .......... .......... 0% 4.03M 2m45s
 1500K .......... .......... .......... .......... .......... 0% 4.08M 2m44s
 1550K .......... .......... .......... .......... .......... 0% 3.49M 2m45s
 1600K .......... .......... .......... .......... .......... 0% 4.10M 2m44s
 1650K .......... .......... .......... .......... .......... 0% 3.88M 2m44s
 1700K .......... .......... .......... .......... .......... 0% 4.00M 2m44s
 1750K .......... .......... .......... .......... .......... 0% 3.45M 2m44s
 1800K .......... .......... .......... .......... .......... 0% 3.97M 2m44s
 1850K .......... .......... .......... .......... .......... 0% 4.08M 2m44s
 1900K .......... .......... .......... .......... .......... 0% 4.18M 2m43s
 1950K .......... .......... .......... .......... .......... 0% 3.54M 2m44s
 2000K .......... .......... .......... .......... .......... 0% 4.15M 2m43s
 2050K .......... .......... .......... .......... .......... 0% 4.13M 2m43s
 2100K .......... .......... .......... .......... .......... 0% 4.13M 2m43s
 2150K .......... .......... .......... .......... .......... 0% 3.40M 2m43s
 2200K .......... .......... .......... .......... .......... 0% 3.95M 2m43s
 2250K .......... .......... .......... .......... .......... 0% 3.89M 2m43s
 2300K .......... .......... .......... .......... .......... 

Next, run the following two cells to copy the Databrick's python packages to the Databrick's File System (DBFS). We'll use these later.

In [5]:
%sh /databricks/python/bin/pip freeze > /tmp/python_packages.txt

In [6]:
%fs cp file:/tmp/python_packages.txt dbfs:/tmp/python_packages.txt

res0: Boolean = true

In the below cell, set the `cluster-name` variable to the name of your cluster, then run the cell. Once complete, click on the cluster's name above (next to the green dot) and restart your cluster. This will install Anaconda, and all of Hail's dependencies. Note that in cluster environments, as long as Python versions between nodes match, it is only necessary to have the Anaconda libraries on the driver.

In [8]:
cluster_name = "qq"
script = """#!/bin/bash
cp /dbfs/tmp/Anaconda3-5.2.0-Linux-x86_64.sh /tmp
sudo bash /tmp/Anaconda3-5.2.0-Linux-x86_64.sh -b -p /anaconda3
mv /databricks/python /databricks/python_old
ln -s /anaconda3 /databricks/python
cp /dbfs/tmp/python_packages.txt /tmp/python_packages.txt
/databricks/python/bin/pip install --upgrade pip
/databricks/python/bin/pip install -r --no-cache-dir /tmp/python_packages.txt
/databricks/python/bin/conda update conda -y
/databricks/python/bin/conda install -y numpy=1.14.0 pandas=0.22.0 matplotlib=2.2.3 seaborn=0.8.1 bokeh=0.12.13
/databricks/python/bin/pip install parsimonious==0.8.0 ipykernel==4.9.0 decorator==4.3.0
"""
dbutils.fs.put("dbfs:/databricks/init/%s/install_conda.sh" % cluster_name, script, True)

Wrote 638 bytes.
Out[11]: True

Once the cluster is restarted, run the below cell. If the output displays
```
3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0]```
Anaconda was installed properly.

In [10]:
%python
import sys
print (sys.version)

3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0]

### Getting Started
Before we can begin, let's import the Hail libraries and initialize Hail.

In [12]:
import hail as hl
import hail.expr.aggregators as agg
hl.init(sc)

Running on Apache Spark version 2.3.1
SparkUI available at http://10.172.234.63:41640
Welcome to
 __ __ <>__
 / /_/ /__ __/ /
 / __ / _ `/ / /
 /_/ /_/\_,_/_/_/ version 0.2-f3019766a4f3
LOGGING: writing to /databricks/driver/hail-20181024-1740-0.2-f3019766a4f3.log

Before using Hail, we import some standard Python libraries for use throughout the notebook.

In [14]:
from pprint import pprint
from bokeh.io import output_notebook, show
from bokeh.layouts import gridplot
from bokeh.models import Span
output_notebook()

###Check tutorial data for download if necessary
This cell downloads the necessary data if it isn’t already present.

In [16]:
hl.utils.get_1kg('/tmp/data')

2018-10-24 17:54:30 Hail: INFO: downloading 1KG VCF ...
 Source: https://storage.googleapis.com/hail-tutorial/1kg.vcf.bgz
2018-10-24 17:54:32 Hail: INFO: importing VCF and writing to matrix table...
2018-10-24 17:54:37 Hail: INFO: Coerced sorted dataset
2018-10-24 17:54:43 Hail: INFO: wrote 10961 items in 16 partitions to /tmp/data/1kg.mt
2018-10-24 17:54:48 Hail: INFO: downloading 1KG annotations ...
 Source: https://storage.googleapis.com/hail-tutorial/1kg_annotations.txt
2018-10-24 17:54:49 Hail: INFO: Done!

### Load data from disk
Hail has its own internal data representation, called a MatrixTable. This is both an on-disk file format and a [Python object](https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable). Here, we read a MatrixTable from disk.

This dataset was created by downsampling a public 1000 genomes VCF to about 50 MB.

In [18]:
mt = hl.read_matrix_table('/tmp/data/1kg.mt')

### Getting to know our data
It’s important to have easy ways to slice, dice, query, and summarize a dataset. Some of these methods are demonstrated below.

The [rows](https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.rows) method can be used to get a table with all the row fields in our MatrixTable.

We can use `rows` along with [select](https://hail.is/docs/0.2/hail.Table.html#hail.Table.select) to pull out 5 variants. The `select` method takes either a string refering to a field name in the table, or a Hail [Expression](https://hail.is/docs/0.2/expr/expression.html#hail.expr.expression.Expression). Here, we leave the arguments blank to keep only the row key fields, `locus` and `alleles`.

Use the `show` method to display the variants.

In [20]:
mt.rows().select().show(5)

+---------------+------------+
 locus | alleles |
+---------------+------------+
 locus<GRCh37> | array<str> |
+---------------+------------+
 1:904165 | ["G","A"] |
 1:909917 | ["G","A"] |
 1:986963 | ["C","T"] |
 1:1563691 | ["T","G"] |
 1:1707740 | ["T","G"] |
+---------------+------------+
showing top 5 rows

Here is how to peek at the first few sample IDs:

In [22]:
mt.s.show(5)

+-----------+
 s |
+-----------+
 str |
+-----------+
 "HG00096" |
 "HG00099" |
 "HG00105" |
 "HG00118" |
 "HG00129" |
+-----------+
showing top 5 rows

To look at the first few genotype calls, we can use [entries](https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.entries) along with `select` and `take`. The `take` method collects the first n rows into a list. Alternatively, we can use the `show` method, which prints the first n rows to the console in a table format.

Try changing `take` to `show` in the cell below.

In [24]:
mt.entry.take(5)

2018-10-24 18:08:26 Hail: WARN: entries(): Resulting entries table is sorted by '(row_key, col_key)'.
 To preserve row-major matrix table order, first unkey columns with 'key_cols_by()'
Out[12]: 
[Struct(GT=0/0, AD=[4, 0], DP=4, GQ=12, PL=[0, 12, 147]),
 Struct(GT=0/0, AD=[8, 0], DP=8, GQ=24, PL=[0, 24, 315]),
 Struct(GT=0/0, AD=[8, 0], DP=8, GQ=23, PL=[0, 23, 230]),
 Struct(GT=0/0, AD=[7, 0], DP=7, GQ=21, PL=[0, 21, 270]),
 Struct(GT=0/0, AD=[5, 0], DP=5, GQ=15, PL=[0, 15, 205])]

### Adding Column Fields
A Hail MatrixTable can have any number of row fields and column fields for storing data associated with each row and column. Annotations are usually a critical part of any genetic study. Column fields are where you’ll store information about sample phenotypes, ancestry, sex, and covariates. Row fields can be used to store information like gene membership and functional impact for use in QC or analysis.

In this tutorial, we demonstrate how to take a text file and use it to annotate the columns in a MatrixTable.

The file provided contains the sample ID, the population and “super-population” designations, the sample sex, and two simulated phenotypes (one binary, one discrete).

This file can be imported into Hail with [import_table](https://hail.is/docs/0.2/methods/impex.html#hail.methods.import_table). This method produces a [Table](https://hail.is/docs/0.2/hail.Table.html#hail.Table) object. Think of this as a Pandas or R dataframe that isn’t limited by the memory on your machine – behind the scenes, it’s distributed with Spark.

In [26]:
table = (hl.import_table('/tmp/data/1kg_annotations.txt', impute=True)
         .key_by('Sample'))

2018-10-24 18:14:28 Hail: INFO: Reading table to impute column types
2018-10-24 18:14:28 Hail: INFO: Finished type imputation
 Loading column 'Sample' as type 'str' (imputed)
 Loading column 'Population' as type 'str' (imputed)
 Loading column 'SuperPopulation' as type 'str' (imputed)
 Loading column 'isFemale' as type 'bool' (imputed)
 Loading column 'PurpleHair' as type 'bool' (imputed)
 Loading column 'CaffeineConsumption' as type 'float64' (imputed)

A good way to peek at the structure of a `Table` is to look at its `schema`.

In [28]:
table.describe()

----------------------------------------
Global fields:
 None
----------------------------------------
Row fields:
 'Sample': str 
 'Population': str 
 'SuperPopulation': str 
 'isFemale': bool 
 'PurpleHair': bool 
 'CaffeineConsumption': float64 
----------------------------------------
Key: ['Sample']
----------------------------------------

To peek at the first few values, use the `show` method:

In [30]:
table.show(width=100)

2018-10-24 18:17:10 Hail: INFO: Coerced sorted dataset
+-----------+------------+-----------------+----------+------------+---------------------+
 Sample | Population | SuperPopulation | isFemale | PurpleHair | CaffeineConsumption |
+-----------+------------+-----------------+----------+------------+---------------------+
 str | str | str | bool | bool | float64 |
+-----------+------------+-----------------+----------+------------+---------------------+
 "HG00096" | "GBR" | "EUR" | false | false | 7.70e+01 |
 "HG00097" | "GBR" | "EUR" | true | true | 6.70e+01 |
 "HG00098" | "GBR" | "EUR" | false | false | 8.30e+01 |
 "HG00099" | "GBR" | "EUR" | true | false | 6.40e+01 |
 "HG00100" | "GBR" | "EUR" | true | false | 5.90e+01 |
 "HG00101" | "GBR" | "EUR" | false | true | 7.70e+01 |
 "HG00102" | "GBR" | "EUR" | true | true | 6.70e+01 |
 "HG00103" | "GBR" | "EUR" | false | true | 7.40e+01 |
 "HG00104" | "GBR" | "EUR" | true | false | 6.60e+01 |
 "HG00105" | "GBR" | "EUR" | false | false | 7.70e+01 |
+-----------+------------+-----------------+----------+------------+---------------------+
showing top 10 rows

Now we’ll use this table to add sample annotations to our dataset, storing the annotations in column fields in our MatrixTable. First, we’ll print the existing column schema:

In [32]:
print(mt.col.dtype)

struct{s: str}

We use the [annotate_cols](https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.annotate_cols) method to join the table with the MatrixTable containing our dataset.

In [34]:
mt = mt.annotate_cols(**table[mt.s])

In [35]:
print(mt.col.dtype)

struct{s: str, Population: str, SuperPopulation: str, isFemale: bool, PurpleHair: bool, CaffeineConsumption: float64}

In [36]:
print(mt.col.dtype.pretty())

struct {
 s: str, 
 Population: str, 
 SuperPopulation: str, 
 isFemale: bool, 
 PurpleHair: bool, 
 CaffeineConsumption: float64
}

###Query functions and the Hail Expression Language
Hail has a number of useful query functions that can be used for gathering statistics on our dataset. These query functions take Hail Expressions as arguments.

We will start by looking at some statistics of the information in our table. The [aggregate](https://hail.is/docs/0.2/hail.Table.html#hail.Table.aggregate) method can be used to aggregate over rows of the table.

`counter` is an aggregation function that counts the number of occurrences of each unique element. We can use this to pull out the population distribution by passing in a Hail Expression for the field that we want to count by.

In [38]:
pprint(table.aggregate(agg.counter(table.SuperPopulation)))

2018-10-24 18:36:03 Hail: INFO: Coerced sorted dataset
{'AFR': 1018, 'AMR': 535, 'EAS': 617, 'EUR': 669, 'SAS': 661}

`stats` is an aggregation function that produces some useful statistics about numeric collections. We can use this to see the distribution of the CaffeineConsumption phenotype.

In [40]:
pprint(table.aggregate(agg.stats(table.CaffeineConsumption)))

2018-10-24 18:36:36 Hail: INFO: Coerced sorted dataset
{'max': 83.0,
 'mean': 67.07600000000005,
 'min': 53.0,
 'n': 3500,
 'stdev': 4.574831269331917,
 'sum': 234766.00000000017}

However, these metrics aren’t perfectly representative of the samples in our dataset. Here’s why:

In [42]:
table.count()

Out[25]: 3500

In [43]:
mt.count_cols()

Out[26]: 284

Since there are fewer samples in our dataset than in the full thousand genomes cohort, we need to look at annotations on the dataset. We can use [aggregate_cols](https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.aggregate_cols) to get the metrics for only the samples in our dataset.

In [45]:
mt.aggregate_cols(agg.counter(mt.SuperPopulation))

2018-10-24 18:38:47 Hail: INFO: Coerced sorted dataset
2018-10-24 18:38:48 Hail: INFO: Coerced sorted dataset
Out[27]: {'AFR': 76, 'EAS': 72, 'AMR': 34, 'SAS': 55, 'EUR': 47}

In [46]:
pprint(mt.aggregate_cols(agg.stats(mt.CaffeineConsumption)))

2018-10-24 18:38:50 Hail: INFO: Coerced sorted dataset
2018-10-24 18:38:50 Hail: INFO: Coerced sorted dataset
{'max': 80.0,
 'mean': 66.80281690140845,
 'min': 57.0,
 'n': 284,
 'stdev': 4.550560152208655,
 'sum': 18972.0}

The functionality demonstrated in the last few cells isn’t anything especially new: it’s certainly not difficult to ask these questions with Pandas or R dataframes, or even Unix tools like `awk`. But Hail can use the same interfaces and query language to analyze collections that are much larger, like the set of variants.

Here we calculate the counts of each of the 12 possible unique SNPs (4 choices for the reference base * 3 choices for the alternate base).

To do this, we need to get the alternate allele of each variant and then count the occurences of each unique ref/alt pair. This can be done with Hail’s `counter` method.

In [48]:
snp_counts = mt.aggregate_rows(agg.counter(hl.Struct(ref=mt.alleles[0], alt=mt.alleles[1])))
pprint(snp_counts)

{Struct(ref=C, alt=A): 496,
 Struct(ref=A, alt=C): 454,
 Struct(ref=C, alt=G): 150,
 Struct(ref=A, alt=T): 76,
 Struct(ref=C, alt=T): 2436,
 Struct(ref=T, alt=A): 79,
 Struct(ref=G, alt=T): 480,
 Struct(ref=T, alt=C): 1879,
 Struct(ref=G, alt=A): 2387,
 Struct(ref=A, alt=G): 1944,
 Struct(ref=G, alt=C): 112,
 Struct(ref=T, alt=G): 468}

We can list the counts in descending order using Python’s Counter class.

In [50]:
from collections import Counter
counts = Counter(snp_counts)
counts.most_common()

Out[30]: 
[(Struct(ref=C, alt=T), 2436),
 (Struct(ref=G, alt=A), 2387),
 (Struct(ref=A, alt=G), 1944),
 (Struct(ref=T, alt=C), 1879),
 (Struct(ref=C, alt=A), 496),
 (Struct(ref=G, alt=T), 480),
 (Struct(ref=T, alt=G), 468),
 (Struct(ref=A, alt=C), 454),
 (Struct(ref=C, alt=G), 150),
 (Struct(ref=G, alt=C), 112),
 (Struct(ref=T, alt=A), 79),
 (Struct(ref=A, alt=T), 76)]

It’s nice to see that we can actually uncover something biological from this small dataset: we see that these frequencies come in pairs. C/T and G/A are actually the same mutation, just viewed from from opposite strands. Likewise, T/A and A/T are the same mutation on opposite strands. There’s a 30x difference between the frequency of C/T and A/T SNPs. Why?

The same Python, R, and Unix tools could do this work as well, but we’re starting to hit a wall - the latest [gnomAD release](http://gnomad.broadinstitute.org/) publishes about 250 million variants, and that won’t fit in memory on a single computer.

What about genotypes? Hail can query the collection of all genotypes in the dataset, and this is getting large even for our tiny dataset. Our 284 samples and 10,000 variants produce 10 million unique genotypes. The gnomAD dataset has about **5 trillion** unique genotypes.

Hail plotting functions allow Hail fields as arguments, so we can pass in the DP field directly here. If the range and bins arguments are not set, this function will compute the range based on minimum and maximum values of the field and use the default 50 bins.

In [52]:
from bokeh.plotting import figure
from bokeh.embed import components, file_html
from bokeh.resources import CDN
p = hl.plot.histogram(mt.DP, range=(0,30), bins=30, title='DP Histogram', legend='DP')
html = file_html(p, CDN, "DP Histogram")
displayHTML(html)


<!DOCTYPE html>
 
 
 
 DP Histogram

### Quality Control
QC is where analysts spend most of their time with sequencing datasets. QC is an iterative process, and is different for every project: there is no “push-button” solution for QC. Each time the Broad collects a new group of samples, it finds new batch effects. However, by practicing open science and discussing the QC process and decisions with others, we can establish a set of best practices as a community.

QC is entirely based on the ability to understand the properties of a dataset. Hail attempts to make this easier by providing the [sample_qc](https://hail.is/docs/0.2/methods/genetics.html#hail.methods.sample_qc) method, which produces a set of useful metrics and stores them in a column field.

In [54]:
mt.col.describe()

--------------------------------------------------------
Type:
 struct {
 s: str, 
 Population: str, 
 SuperPopulation: str, 
 isFemale: bool, 
 PurpleHair: bool, 
 CaffeineConsumption: float64
 }
--------------------------------------------------------
Source:
 <hail.matrixtable.MatrixTable object at 0x7f3aca240fd0>
Index:
 ['column']
--------------------------------------------------------

In [55]:
mt = hl.sample_qc(mt)

In [56]:
mt.col.describe()

--------------------------------------------------------
Type:
 struct {
 s: str, 
 Population: str, 
 SuperPopulation: str, 
 isFemale: bool, 
 PurpleHair: bool, 
 CaffeineConsumption: float64, 
 sample_qc: struct {
 dp_stats: struct {
 mean: float64, 
 stdev: float64, 
 min: float64, 
 max: float64
 }, 
 gq_stats: struct {
 mean: float64, 
 stdev: float64, 
 min: float64, 
 max: float64
 }, 
 call_rate: float64, 
 n_called: int64, 
 n_not_called: int64, 
 n_hom_ref: int64, 
 n_het: int64, 
 n_hom_var: int64, 
 n_non_ref: int64, 
 n_singleton: int64, 
 n_snp: int64, 
 n_insertion: int64, 
 n_deletion: int64, 
 n_transition: int64, 
 n_transversion: int64, 
 n_star: int64, 
 r_ti_tv: float64, 
 r_het_hom_var: float64, 
 r_insertion_deletion: float64
 }
 }
--------------------------------------------------------
Source:
 <hail.matrixtable.MatrixTable object at 0x7f3ab1556f98>
Index:
 ['column']
--------------------------------------------------------

Plotting the QC metrics is a good place to start.

In [58]:
p = hl.plot.histogram(mt.sample_qc.call_rate, range=(.88,1), legend='Call Rate')
html = file_html(p, CDN, "Call Rate")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Call Rate

In [59]:
p = hl.plot.histogram(mt.sample_qc.gq_stats.mean, range=(10,70), legend='Mean Sample GQ')
html = file_html(p, CDN, "Mean Sample GQ")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Mean Sample GQ

Often, these metrics are correlated.

In [61]:
p = hl.plot.scatter(mt.sample_qc.dp_stats.mean, mt.sample_qc.call_rate, xlabel='Mean DP', ylabel='Call Rate')
html = file_html(p, CDN, "Mean DP vs Call Rate")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Mean DP vs Call Rate

Removing outliers from the dataset will generally improve association results. We can draw lines on the above plot to indicate outlier cuts.

In [63]:
p.renderers.extend(
    [Span(location=0.97, dimension='width', line_color='black', line_width=1),
     Span(location=4, dimension='height', line_color='black', line_width=1)])
html = file_html(p, CDN, "Mean DP vs Call Rate")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Mean DP vs Call Rate

We’ll want to remove all samples that fall in the bottom left quadrant.

It’s easy to filter when we’ve got the cutoff values decided:

In [65]:
mt = mt.filter_cols((mt.sample_qc.dp_stats.mean >= 4) & (mt.sample_qc.call_rate >= 0.97))
print('After filter, %d/284 samples remain.' % mt.count_cols())

2018-10-24 18:52:43 Hail: INFO: Coerced sorted dataset
After filter, 250/284 samples remain.

Next is genotype QC. To start, we’ll print the post-sample-QC call rate. It’s actually gone up since the initial summary - dropping low-quality samples disproportionally removed missing genotypes.

Import the `hail.expr.functions` module.

In [67]:
call_rate = mt.aggregate_entries(agg.fraction(hl.is_defined(mt.GT)))
print('pre QC call rate is %.3f' % call_rate)

pre QC call rate is 0.992

It’s a good idea to filter out genotypes where the reads aren’t where they should be: if we find a genotype called homozygous reference with >10% alternate reads, a genotype called homozygous alternate with >10% reference reads, or a genotype called heterozygote without a ref / alt balance near 1:1, it is likely to be an error.

In [69]:
ab = mt.AD[1] / hl.sum(mt.AD)

filter_condition_ab = ((mt.GT.is_hom_ref() & (ab <= 0.1)) |
                        (mt.GT.is_het() & (ab >= 0.25) & (ab <= 0.75)) |
                        (mt.GT.is_hom_var() & (ab >= 0.9)))

mt = mt.filter_entries(filter_condition_ab)

In [70]:
post_qc_call_rate = mt.aggregate_entries(agg.fraction(hl.is_defined(mt.GT)))
print('post QC call rate is %.3f' % post_qc_call_rate)

post QC call rate is 0.955

Variant QC is a bit more of the same: we can use the [variant_qc](https://hail.is/docs/0.2/methods/genetics.html?highlight=variant%20qc#hail.methods.variant_qc) method to produce a variety of useful statistics, plot them, and filter.

In [72]:
mt.row.describe()

--------------------------------------------------------
Type:
 struct {
 locus: locus<GRCh37>, 
 alleles: array<str>, 
 rsid: str, 
 qual: float64, 
 filters: set<str>, 
 info: struct {
 AC: array<int32>, 
 AF: array<float64>, 
 AN: int32, 
 BaseQRankSum: float64, 
 ClippingRankSum: float64, 
 DP: int32, 
 DS: bool, 
 FS: float64, 
 HaplotypeScore: float64, 
 InbreedingCoeff: float64, 
 MLEAC: array<int32>, 
 MLEAF: array<float64>, 
 MQ: float64, 
 MQ0: int32, 
 MQRankSum: float64, 
 QD: float64, 
 ReadPosRankSum: float64, 
 set: str
 }
 }
--------------------------------------------------------
Source:
 <hail.matrixtable.MatrixTable object at 0x7f3ab1543908>
Index:
 ['row']
--------------------------------------------------------

In [73]:
mt = hl.variant_qc(mt).cache()

2018-10-24 18:55:50 Hail: INFO: Coerced sorted dataset
2018-10-24 18:55:54 Hail: INFO: Coerced sorted dataset

In [74]:
mt.row.describe()

--------------------------------------------------------
Type:
 struct {
 locus: locus<GRCh37>, 
 alleles: array<str>, 
 rsid: str, 
 qual: float64, 
 filters: set<str>, 
 info: struct {
 AC: array<int32>, 
 AF: array<float64>, 
 AN: int32, 
 BaseQRankSum: float64, 
 ClippingRankSum: float64, 
 DP: int32, 
 DS: bool, 
 FS: float64, 
 HaplotypeScore: float64, 
 InbreedingCoeff: float64, 
 MLEAC: array<int32>, 
 MLEAF: array<float64>, 
 MQ: float64, 
 MQ0: int32, 
 MQRankSum: float64, 
 QD: float64, 
 ReadPosRankSum: float64, 
 set: str
 }, 
 variant_qc: struct {
 AC: array<int32>, 
 AF: array<float64>, 
 AN: int32, 
 homozygote_count: array<int32>, 
 dp_stats: struct {
 mean: float64, 
 stdev: float64, 
 min: float64, 
 max: float64
 }, 
 gq_stats: struct {
 mean: float64, 
 stdev: float64, 
 min: float64, 
 max: float64
 }, 
 n_called: int64, 
 n_not_called: int64, 
 call_rate: float32, 
 n_het: int64, 
 n_non_ref: int64, 
 het_freq_hwe: float64, 
 p_value_hwe: float64
 }
 }
--------------------------------------------------------
Source:
 <hail.matrixtable.MatrixTable object at 0x7f3ab14d4828>
Index:
 ['row']
--------------------------------------------------------

In [75]:
qc_table = mt.rows()
qc_table = qc_table.select(gq_mean = qc_table.variant_qc.gq_stats.mean,
                           call_rate = qc_table.variant_qc.call_rate,
                           p_value_hwe = qc_table.variant_qc.p_value_hwe,
                           AAF = qc_table.variant_qc.AF[1],)

p1 = hl.plot.histogram(qc_table.gq_mean, range=(10,80), legend='Variant Mean GQ')

p2 = hl.plot.histogram(qc_table.AAF, range=(0,1), legend='Alternate Allele Frequency')

p3 = hl.plot.histogram(qc_table.call_rate, range=(.5,1), legend='Variant Call Rate')

p4 = hl.plot.histogram(qc_table.p_value_hwe, range=(0,1), legend='Hardy-Weinberg Equilibrium p-value')

g = gridplot([[p1, p2], [p3, p4]])

html = file_html(g, CDN, "Mean DP vs Call Rate")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Mean DP vs Call Rate

These statistics actually look pretty good: we don’t need to filter this dataset. Most datasets require thoughtful quality control, though. The [filter_rows])(https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable.filter_rows) method can help!

### Let's do a GWAS
First, we need to restrict to variants that are:

* common (we’ll use a cutoff of 1%)
* uncorrelated (not in linkage disequilibrium)

In [78]:
common_mt = mt.filter_rows(mt.variant_qc.AF[1] > 0.01)

In [79]:
print('Samples: %d  Variants: %d' % (common_mt.count_cols(), common_mt.count_rows()))

Samples: 250 Variants: 9177

These filters removed about 15% of sites (we started with a bit over 10,000). This is NOT representative of most sequencing datasets! We have already downsampled the full thousand genomes dataset to include more common variants than we’d expect by chance.

In Hail, the association tests accept column fields for the sample phenotype and covariates. Since we’ve already got our phenotype of interest (caffeine consumption) in the dataset, we are good to go:

In [81]:
gwas = hl.linear_regression_rows(y=common_mt.CaffeineConsumption, x=common_mt.GT.n_alt_alleles(), covariates=[1.0])
gwas.row.describe()

2018-10-24 19:02:40 Hail: INFO: linear_regression_rows: running on 250 samples for 1 response variable y,
 with input variable x, and 1 additional covariate...
--------------------------------------------------------
Type:
 struct {
 locus: locus<GRCh37>, 
 alleles: array<str>, 
 n: int32, 
 sum_x: float64, 
 y_transpose_x: float64, 
 beta: float64, 
 standard_error: float64, 
 t_stat: float64, 
 p_value: float64
 }
--------------------------------------------------------
Source:
 <hail.table.Table object at 0x7f3ab141abe0>
Index:
 ['row']
--------------------------------------------------------

Looking at the bottom of the above printout, you can see the linear regression adds new row fields for the beta, standard error, t-statistic, and p-value.

Hail makes it easy to make a [Q-Q (quantile-quantile) plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot).

In [83]:
p = hl.plot.qq(gwas.p_value)

html = file_html(p, CDN, "GWAS p-value")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 GWAS p-value

###Confounded!
The observed p-values drift away from the expectation immediately. Either every SNP in our dataset is causally linked to caffeine consumption (unlikely), or there’s a confounder.

We didn’t tell you, but sample ancestry was actually used to simulate this phenotype. This leads to a [stratified](https://en.wikipedia.org/wiki/Population_stratification) distribution of the phenotype. The solution is to include ancestry as a covariate in our regression.

The [linear_regression](https://hail.is/docs/0.2/methods/stats.html#hail.methods.linear_regression) method can also take column fields to use as covariates. We already annotated our samples with reported ancestry, but it is good to be skeptical of these labels due to human error. Genomes don’t have that problem! Instead of using reported ancestry, we will use genetic ancestry by including computed principal components in our model.

The [pca](https://hail.is/docs/0.2/methods/stats.html#hail.methods.pca) method produces eigenvalues as a list and sample PCs as a Table, and can also produce variant loadings when asked. The [hwe_normalized_pca](https://hail.is/docs/0.2/methods/genetics.html#hail.methods.hwe_normalized_pca) method does the same, using HWE-normalized genotypes for the PCA.

In [85]:
pca_eigenvalues, pca_scores, _ = hl.hwe_normalized_pca(common_mt.GT)

2018-10-24 19:07:31 Hail: INFO: hwe_normalized_pca: running PCA using 9169 variants.
2018-10-24 19:07:34 Hail: INFO: pca: running PCA with 10 components...
2018-10-24 19:10:36 Hail: INFO: Coerced sorted dataset

In [86]:
pprint(pca_eigenvalues)

[23.64283149708918,
 11.390991302730127,
 3.628037491657114,
 2.755633938634912,
 1.5876756051186907,
 1.510530614633275,
 1.4849653335891342,
 1.4579822471398736,
 1.4425262940776513,
 1.4220076265223383]

In [87]:
p = hl.plot.scatter(pca_scores.scores[0], pca_scores.scores[1],
                    label=common_mt.cols()[pca_scores.s].SuperPopulation,
                    title='PCA', xlabel='PC1', ylabel='PC2')
html = file_html(p, CDN, "PCA")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 PCA

Now we can rerun our linear regression, controlling for sample sex and the first few principal components. We’ll do this with input variable the number of alternate alleles as before, and again with input variable the genotype dosage derived from the PL field.

In [89]:
cmt = common_mt.annotate_cols(pca = pca_scores[common_mt.s])

linear_regression_results = hl.linear_regression_rows(
    y=cmt.CaffeineConsumption, x=cmt.GT.n_alt_alleles(),
    covariates=[1.0, cmt.isFemale, cmt.pca.scores[0], cmt.pca.scores[1], cmt.pca.scores[2]])
  
p = hl.plot.qq(linear_regression_results.p_value)
html = file_html(p, CDN, "Q-Q Plot")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Q-Q Plot

We can also use the genotype dosage (the probability-weighted expected number of alternate alleles) instead of the hard call in our regression

In [91]:
def plDosage(pl):
    linear_scaled = pl.map(lambda x: 10 ** (-x / 10))
    pl_sum = hl.sum(linear_scaled)
    normed = linear_scaled / pl_sum
    return 1 * normed[1] + 2 * normed[2]

In [92]:
linear_regression_results = hl.linear_regression_rows(
    y=cmt.CaffeineConsumption, x=plDosage(cmt.PL),
    covariates=[1.0, cmt.isFemale, cmt.pca.scores[0], cmt.pca.scores[1], cmt.pca.scores[2]])

2018-10-24 19:16:11 Hail: INFO: linear_regression_rows: running on 250 samples for 1 response variable y,
 with input variable x, and 5 additional covariates...

In [93]:
p = hl.plot.qq(linear_regression_results.p_value)
html = file_html(p, CDN, "Q-Q Plot")
displayHTML(html)

<!DOCTYPE html>
 
 
 
 Q-Q Plot

That’s more like it! We may not be publishing ten new coffee-drinking loci in Nature, but we shouldn’t expect to find anything but the strongest signals from a dataset of 284 individuals anyway.

###Rare variant analysis
Here we’ll demonstrate how one can use the expression language to group and count by any arbitrary properties in row and column fields. Hail also implements the sequence kernel association test (SKAT).

In [96]:
entries = mt.entries()
results = (entries.group_by(pop = entries.SuperPopulation, chromosome = entries.locus.contig)
      .aggregate(n_het = agg.count_where(entries.GT.is_het())))

In [97]:
results.show()

2018-10-24 19:18:22 Hail: INFO: Ordering unsorted dataset with network shuffle
+-------+------------+-------+
 pop | chromosome | n_het |
+-------+------------+-------+
 str | str | int64 |
+-------+------------+-------+
 "AFR" | "1" | 12583 |
 "AFR" | "10" | 7950 |
 "AFR" | "11" | 7872 |
 "AFR" | "12" | 7876 |
 "AFR" | "13" | 5251 |
 "AFR" | "14" | 4866 |
 "AFR" | "15" | 4459 |
 "AFR" | "16" | 5306 |
 "AFR" | "17" | 4428 |
 "AFR" | "18" | 4583 |
+-------+------------+-------+
showing top 10 rows

What if we want to group by minor allele frequency bin and hair color, and calculate the mean GQ?

In [99]:
entries = entries.annotate(maf_bin = hl.cond(entries.info.AF[0]<0.01, "< 1%",
                             hl.cond(entries.info.AF[0]<0.05, "1%-5%", ">5%")))

results2 = (entries.group_by(af_bin = entries.maf_bin, purple_hair = entries.PurpleHair)
      .aggregate(mean_gq = agg.stats(entries.GQ).mean,
                 mean_dp = agg.stats(entries.DP).mean))

In [100]:
results2.show()

2018-10-24 19:19:19 Hail: INFO: Coerced sorted dataset
2018-10-24 19:19:19 Hail: INFO: Coerced dataset with out-of-order partitions.
+---------+-------------+----------+----------+
 af_bin | purple_hair | mean_gq | mean_dp |
+---------+-------------+----------+----------+
 str | bool | float64 | float64 |
+---------+-------------+----------+----------+
 "1%-5%" | false | 2.43e+01 | 7.40e+00 |
 "1%-5%" | true | 2.42e+01 | 7.43e+00 |
 "< 1%" | false | 2.31e+01 | 7.50e+00 |
 "< 1%" | true | 2.32e+01 | 7.55e+00 |
 ">5%" | false | 3.61e+01 | 7.52e+00 |
 ">5%" | true | 3.65e+01 | 7.57e+00 |
+---------+-------------+----------+----------+

We’ve shown that it’s easy to aggregate by a couple of arbitrary statistics. This specific examples may not provide especially useful pieces of information, but this same pattern can be used to detect effects of rare variation:

* Count the number of heterozygous genotypes per gene by functional category (synonymous, missense, or loss-of-function) to estimate per-gene functional constraint* 
* Count the number of singleton loss-of-function mutations per gene in cases and controls to detect genes involved in disease

###Epilogue
Congrats! You’ve reached the end of the first tutorial. To learn more about Hail’s API and functionality, take a look at the other tutorials. You can check out the [Python API](https://hail.is/docs/0.2/api.html#python-api) for documentation on additional Hail functions. If you use Hail for your own science, we’d love to hear from you on [Zulip chat](https://hail.zulipchat.com/) or the [discussion forum](http://discuss.hail.is/).

For reference, here’s the full workflow to all tutorial endpoints combined into one cell. It may take a minute! It’s doing a lot of work.

In [103]:
table = hl.import_table('/tmp/data/1kg_annotations.txt', impute=True).key_by('Sample')

mt = hl.read_matrix_table('/tmp/data/1kg.mt')
mt = mt.annotate_cols(**table[mt.s])
mt = hl.sample_qc(mt)
mt = mt.filter_cols((mt.sample_qc.dp_stats.mean >= 4) & (mt.sample_qc.call_rate >= 0.97))
ab = mt.AD[1] / hl.sum(mt.AD)
filter_condition_ab = ((mt.GT.is_hom_ref() & (ab <= 0.1)) |
                        (mt.GT.is_het() & (ab >= 0.25) & (ab <= 0.75)) |
                        (mt.GT.is_hom_var() & (ab >= 0.9)))
mt = mt.filter_entries(filter_condition_ab)
mt = hl.variant_qc(mt)
common_mt = mt.filter_rows(mt.variant_qc.AF[1] > 0.01)

pca_eigenvalues, pca_scores, _ = hl.hwe_normalized_pca(common_mt.GT)

cmt = common_mt.annotate_cols(pca = pca_scores[common_mt.s])
linear_regression_results = hl.linear_regression_rows(
    y=cmt.CaffeineConsumption, x=cmt.GT.n_alt_alleles(),
    covariates=[1.0, cmt.isFemale, cmt.pca.scores[0], cmt.pca.scores[1], cmt.pca.scores[2]])
pvals = linear_regression_results.p_value.collect()